src/Compilers/CSharp/Portable/CodeGen/EmitArrayInitializer.cs

```diff
@@ -445,54 +501,192 @@ private bool TryEmitReadonlySpanAsBlobWrapper(NamedTypeSymbol spanType, BoundExp
             }
 
-            if (!inPlace && !used)
+            if (inPlaceTarget is null && !used)
             {
-                // emitting a value that no one will see
+                // The caller has specified that we're creating a ReadOnlySpan expression that won't be used.
+                // We needn't emit anything.
                 return true;
             }
 
             if (elementCount == 0)
             {
-                if (inPlace)
+                // The span is empty.  Optimize away the array.  This works regardless of the size of the type.
+                // (We could optimize this even for non-primitives, but it's not currently worthwhile.)
+
+                // If this is in-place initialization, call the default ctor.
+                if (inPlaceTarget is not null)
                 {
+                    EmitAddress(inPlaceTarget, Binder.AddressKind.Writeable);
                     _builder.EmitOpCode(ILOpCode.Initobj);
                     EmitSymbolToken(spanType, wrappedExpression.Syntax);
+                    if (used)
+                    {
+                        EmitExpression(inPlaceTarget, used: true);
+                    }
                 }
                 else
                 {
+                    // Otherwise, assign it to a default value / empty span.
                     EmitDefaultValue(spanType, used, wrappedExpression.Syntax);
                 }
+                return true;
             }
-            else
+
+            if (IsPeVerifyCompatEnabled())
+            {
+                // After this point, we're emitting code that may cause PEVerify to warn, so stop if PEVerify compat is enabled.
+                return false;
+            }
+
+            if (specialElementType.SizeInBytes() == 1)
             {
-                if (IsPeVerifyCompatEnabled())
+                // We're dealing with a ReadOnlySpan<byte/sbyte/bool>. We can optimize this on all target platforms,
+                // whether the initialization is in-place or not.
+
+                if (inPlaceTarget is not null)
                 {
-                    return false;
+                    EmitAddress(inPlaceTarget, Binder.AddressKind.Writeable);
                 }
 
-                _builder.EmitArrayBlockFieldRef(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag!);
+                _builder.EmitArrayBlockFieldRef(data, wrappedExpression.Syntax, _diagnostics.DiagnosticBag);
                 _builder.EmitIntConstant(lengthForConstructor);
 
-                if (inPlace)
+                if (inPlaceTarget is not null)
                 {
-                    // consumes target ref, data ptr and size, pushes nothing
+                    // Consumes target ref, data ptr and size, pushes nothing.
                     _builder.EmitOpCode(ILOpCode.Call, stackAdjustment: -3);
+                    if (used)
+                    {
+                        EmitExpression(inPlaceTarget, used: true);
```

Comment:
>EmitExpression(inPlaceTarget, used: true); [](http://example.com/codeflow?start=24&length=42)

Should we be emitting constructor token before doing this?

Reply:
I can't actually get this code to execute.  Even in main, if I run all ~5000 of the code gen tests, this line:
https://github.com/dotnet/roslyn/blob/eebc8b739b34875bd3a2f6c28a3fe2c327ca6405/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L2254
is never hit.  I can either leave the code as I have it, which is effectively what was there before (the line just moved into this method), or add a throw Unreachable / assert here.  What would you prefer?

Reply:
>>Should we be emitting constructor token before doing this?

Regardless of question how to hit this code path. What is the answer to the question? Shouldn't the call include constructor's token?

Reply:
The token is emitted below just after the call opcode:
https://github.com/dotnet/roslyn/pull/61414/files/c2797454921052f2df2fb3479f8116b4a6d88fd0#diff-1e1b32781086289e35025f3ef987f012bfe6f3fec3a25312e5f3566a977e9d98R573

Reply:
> the line just moved into this method

This is not how I see it, It looks to me that the line git injected in the middle of IL instruction. We haven finished emitting IL instruction for the call

Reply:
>> the line just moved into this method

> This is not how I see it

My point was that this line was never hit in main in its original location and is still never being hit.  I've deleted it in my local changes.

Reply:
> That's how it is today in main, too:

I do not see anything emitted between the op code and the method token in main

Reply:
Aleksey, I just said "My comments about the call and token are based on this if (used) block simply not existing, because it doesn't anymore.".

Can you please wait to continue asking questions about this until I push up my revision?

Thanks.

Reply:
> I just said "My comments about the call and token are based on this if (used) block simply not existing, because it doesn't anymore.".

Lets mot rush with removing the block. I'd prefer us spending some time trying to get this code path hit. The fix, I think is obvious - don't emit things in the middle of IL command, finish it and then emit the value

Reply:
> I'd prefer us spending some time trying to get this code path hit.

In main today, this line is never hit by any of the ~5000 code gen tests:
https://github.com/dotnet/roslyn/blob/eebc8b739b34875bd3a2f6c28a3fe2c327ca6405/src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs#L2254
In main, what code will cause that to be hit?

Reply:
We need to test this scenario anyway, whether the optimization is done or not.

Reply:
> We need to test this scenario anyway, whether the optimization is done or not.

What scenario?  I'm telling you I do not see how to hit that line.  You asked me to ensure all branches were covered, and I'm telling you I don't know how to cover an existing branch in the existing code.  Whether or not I moved it to the right exact location in this method doesn't change the fact that it's never hit by any existing test and I don't know how it's possible to get it to trigger.

So, I'm asking you for help: what code will test "this scenario"?  Forget this PR: what code will test that line in main?

Reply:
Thanks for helping me offline to come up with a way to hit it.